src/main/java/org/apache/tika/parser/external/WaitingExternalParser.java

```diff
@@ -15,48 +13,391 @@
+    public void setCommand(String... command) {
+        this.command = command;
+    }
+    
+    
+    public Map<Pattern,String> getMetadataExtractionPatterns() {
+       return metadataPatterns;
+    }
+    
+    /**
+     * Sets the map of regular expression patterns and Metadata
+     *  keys. Any matching patterns will have the matching
+     *  metadata entries set.
+     * Set this to null to disable Metadata extraction.
+     */
+    public void setMetadataExtractionPatterns(Map<Pattern,String> patterns) {
+       this.metadataPatterns = patterns;
+    }
     
-    private static final long PATTERN_MATCHING_DELAY_MS = 20;
 
-    @Override
+    /**
+     * Executes the configured external command and passes the given document
+     *  stream as a simple XHTML document to the given SAX content handler.
+     * Metadata is only extracted if {@link #setMetadataExtractionPatterns(Map)}
+     *  has been called to set patterns.
+     */
     public void parse(
             InputStream stream, ContentHandler handler,
             Metadata metadata, ParseContext context)
             throws IOException, SAXException, TikaException {
-        super.parse(stream, handler, metadata, context);
-        // Pattern checking is done in a new Thread which must be allowed to finish
+        XHTMLContentHandler xhtml =
+            new XHTMLContentHandler(handler, metadata);
+
+        TemporaryResources tmp = new TemporaryResources();
         try {
-            Thread.sleep(PATTERN_MATCHING_DELAY_MS);
-        } catch (InterruptedException e) {
-            // We were asked to stop
+            parse(TikaInputStream.get(stream, tmp),
+                    xhtml, metadata, tmp);
+        } finally {
+            tmp.dispose();
+        }
+    }
+
+    private void parse(
+            TikaInputStream stream, XHTMLContentHandler xhtml,
+            Metadata metadata, TemporaryResources tmp)
+            throws IOException, SAXException, TikaException {
+        boolean inputToStdIn = true;
+        boolean outputFromStdOut = true;
+        boolean hasPatterns = (metadataPatterns != null && !metadataPatterns.isEmpty());
+
+        File output = null;
+
+        // Build our command
+        String[] cmd = new String[command.length];
+        System.arraycopy(command, 0, cmd, 0, command.length);
+        for(int i=0; i<cmd.length; i++) {
+           if(cmd[i].indexOf(INPUT_FILE_TOKEN) != -1) {
+              cmd[i] = cmd[i].replace(INPUT_FILE_TOKEN, stream.getFile().getPath());
+              inputToStdIn = false;
+           }
+           if(cmd[i].indexOf(OUTPUT_FILE_TOKEN) != -1) {
+              output = tmp.createTemporaryFile();
+              outputFromStdOut = false;
+           }
         }
+
+        // Execute
+        Process process;
+        if(cmd.length == 1) {
+           process = Runtime.getRuntime().exec( cmd[0] );
+        } else {
+           process = Runtime.getRuntime().exec( cmd );
+        }
+
+        Parser stdoutParser = null;
```

Comment:
Calling these parsers (within a class called *Parser) is going to lead to confusion.  Perhaps `stdoutStreamParser` and `stderrStreamParser`.